tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -131,31 +152,149 @@ func (a *AwsProvider) CreateWindowsVM() (rerr error) {
 	instance, err := a.createInstance(a.imageID, a.instanceType, a.sshKey, networkInterface, workerIAM, userDataWinrm)
 
 	if err != nil {
-		return err
+		return nil, err
 	}
 	instanceID := *instance.InstanceId
 
 	// Wait until instance is running and associate a unique name tag to the created instance.
 	err = a.waitUntilInstanceRunning(instanceID)
 	if err != nil {
-		return fmt.Errorf("failed to wait till instance is running, %v", err)
+		return nil, fmt.Errorf("failed to wait till instance is running, %v", err)
 	}
 	_, err = a.createInstanceNameTag(instance, infraID)
 	if err != nil {
 		log.V(0).Info(fmt.Sprintf("failed to assign name for instance: %s, %v", instanceID, err))
 	}
 
-	err = resource.AppendInstallerInfo([]string{instanceID}, []string{}, a.resourceTrackerDir)
+	// Get the public IP
+	publicIPAddress, err := a.GetPublicIP(instanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the decrypted password
+	decryptedPassword, err := a.GetPassword(instanceID)
 	if err != nil {
-		return fmt.Errorf("failed to record instance ID to file at '%s',instance will not be able to be deleted, "+
+		return nil, fmt.Errorf("error with instance creation %v", err)
+	}
+
+	// Build new credentials structure to be used by other actors
+	credentials = types.NewCredentials(instanceID, publicIPAddress, decryptedPassword)
+	err = resource.AppendInstallerInfo([]string{instanceID}, []string{}, a.resourceTrackerDir)
+	if err != nil || credentials == nil {
```

Comment:
It seems like this should be separated into two error blocks.  Credentials being nil has no effect on writing the details to the file. So if writing was successful, and we could not get the credentials, the error message will be false.

Reply:
While I am fine splitting it, I think creating the `windows-installer.json` without the password being available is useless. If it confuses, I'll remove the check here and have the check in entity which is calling the `CreateVM` method.